modules/message-lane/src/inbound_lane.rs

```diff
@@ -0,0 +1,236 @@
+
+use bp_message_lane::{InboundLaneData, LaneId, Message, MessageKey, MessageNonce, MessageResult, OnMessageReceived};
+
+/// Inbound lane storage.
+pub trait InboundLaneStorage {
+	/// Message payload.
+	type Payload;
+
+	/// Lane id.
+	fn id(&self) -> LaneId;
+	/// Get lane data from the storage.
+	fn data(&self) -> InboundLaneData;
+	/// Update lane data in the storage.
+	fn set_data(&mut self, data: InboundLaneData);
+	/// Returns saved inbound message payload.
+	fn message(&self, nonce: &MessageNonce) -> Option<Self::Payload>;
+	/// Save inbound message in the storage.
+	fn save_message(&mut self, nonce: MessageNonce, payload: Self::Payload);
+	/// Remove inbound message from the storage.
+	fn remove_message(&mut self, nonce: &MessageNonce);
+}
+
+/// Inbound messages lane.
+pub struct InboundLane<Storage> {
+	storage: Storage,
+}
+
+impl<Storage: InboundLaneStorage> InboundLane<Storage> {
+	/// Create new inbound lane backed by given storage.
+	pub fn new(storage: Storage) -> Self {
+		InboundLane { storage }
+	}
+
+	/// Receive new message.
+	pub fn receive_message(
+		&mut self,
+		nonce: MessageNonce,
+		payload: Storage::Payload,
+		processor: &mut impl OnMessageReceived<Storage::Payload>,
+	) -> bool {
+		let mut data = self.storage.data();
+		let is_correct_message = nonce == data.latest_received_nonce + 1;
+		if !is_correct_message {
+			return false;
+		}
+
+		let is_process_required = is_correct_message && data.oldest_unprocessed_nonce == nonce;
+		data.latest_received_nonce = nonce;
+		self.storage.set_data(data);
+
+		let payload_to_save = match is_process_required {
+			true => {
+				let message = Message {
+					key: MessageKey {
+						lane_id: self.storage.id(),
+						nonce,
+					},
+					payload,
+				};
+				match processor.on_message_received(message) {
+					MessageResult::Processed => None,
+					MessageResult::NotProcessed(message) => Some(message.payload),
+				}
+			}
+			false => Some(payload),
+		};
+
+		if let Some(payload_to_save) = payload_to_save {
+			self.storage.save_message(nonce, payload_to_save);
+		}
+
+		true
+	}
+
+	/// Process stored lane messages.
+	///
+	/// Stops processing either when all messages are processed, or when processor returns
+	/// MessageResult::NotProcessed.
+	pub fn process_messages(&mut self, processor: &mut impl OnMessageReceived<Storage::Payload>) {
+		let mut anything_processed = false;
+		let mut data = self.storage.data();
+		while data.oldest_unprocessed_nonce <= data.latest_received_nonce {
+			let nonce = data.oldest_unprocessed_nonce;
+			let payload = self
+				.storage
+				.message(&nonce)
+				.expect("message is referenced by lane; referenced message is not pruned; qed");
+			let message = Message {
+				key: MessageKey {
+					lane_id: self.storage.id(),
+					nonce,
+				},
+				payload,
+			};
+
+			let process_result = processor.on_message_received(message);
+			if let MessageResult::NotProcessed(_) = process_result {
+				break;
+			}
```

Comment:
Would it be possible for us to get stuck with a bad message?

Let's say we have messages A, and B in storage pending processing. We'll process them A->B. Since A is the oldest, it'll get processed first, next time we try and process B but fail for some reason.

Now a new message, C comes in. Next time we try and process messages we'll start at the oldest, B, but we'll continue to get `NotProcessed` and never make our way to C - effectively blocking the lane.

Reply:
Sure - bad processor/handler is able to block lane. If handler would be unable to handle message forever, the lane would stuck. The other option is to force it to handle message as it appears, but it can be impossible because of block time/weight. So it is up to handler to decide when and how to process messages.

Reply:
I think it's fine for the processor to be able to decide when to handle messages, but I'm still wondering if there's some way to get unstuck. It seems like a pretty simple attack to spam the chain with invalid messages just to block lanes.

Would it be possible to send a different valid message at the same nonce which can take priority?

Reply:
I the message is invalid, then the handler needs to fail the message. I.e. just ignore + return `Processed`. That's the most straightforward way imo :)

Reply:
Yes. There is an issue with a seemingly valid message that doesn't fit the block size - i.e. the handler (message processor) MUST detect that and fail the message right away. Might be worth adding to the documentation of `OnMessageReceived`.

Reply:
Updated docs